### PR TITLE
HPCC-27988 Allow Singleton initialization to null

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jmutex.hpp [hunk 1]
```diff
@@ -1002,22 +1002,37 @@ template <typename X>
 class Singleton
 {
 public:
-    template <typename FUNC> X * query(FUNC factory) { return querySingleton(singleton, cs, factory); }
+    template <typename FUNC> X * query(FUNC factory)
+    {
```
Comment (jakesmith) [inline system/jlib/jmutex.hpp:1006]: @ghalliday - considered adding a querySingleton() variant (that took an initializer atomic too), but I don't think it's worth it, or worth changing the existing direct uses of querySingleton (all in deftype.cpp)

File: system/jlib/jmutex.hpp [hunk 2]
```diff
@@ -1002,22 +1002,37 @@ template <typename X>
...
+        if (initialized.load(std::memory_order_acquire))
+            return singleton.load(std::memory_order_acquire); // avoid crit
+        CriticalBlock block(cs);
+        X * value = singleton.load(std::memory_order_acquire); // reload in case another thread got here first
+        if (value) // implies initialized=true also
```
Comment (ghalliday) [inline system/jlib/jmutex.hpp:1011]: should be checking initialised rather than value - in case the factory is expensive, or has side-effects.

**Review by ghalliday — COMMENTED** [review]
@jakesmith one minor comment

File: system/jlib/jmutex.hpp (hunk 2 above)
Comment (jakesmith) [reply system/jlib/jmutex.hpp]: changed

**Review by ghalliday — APPROVED** [review]
Looks good, please squash.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**